### ISSUE TITLE
[v2.0] Should we extend X584 comments with input directive?

### ISSUE THREAD

**Issue by danil-kondr2016** [kodemeister/X584#11]
This may be handy for end users since they will be able to specify input values without entering them each time when program is running.

But there are two major cons:
1. Extending comment format will create new format derived from X584 which is not backward compatible with previous one;
2. Comments are used not only for specifying conditional jumps inside the code but primarily **for commenting**. For example, one's write code which loads floating-point number represented with two integers (mantissa and power (or exponent)) like that:
  ```
  РОН0 := ШИНвх   ; mA
  РОН1 := ШИНвх   ; pA
  РОН2 := ШИНвх   ; mB
  РОН3 := ШИНвх   ; pB
  ```
  If this extension will be implemented we won't be able to properly comment these sections: for example, we can't add extra comments inside conditional ones (like `if FLAG then TRUE_ADDR [else FALSE_ADDR]`)

**Comment by kodemeister** [comment]
Hey Danila,

> Extending comment format will create new format derived from X584 which is not backward compatible with previous one;

Well, not really. Adding new keywords like `input` doesn't mean creating a new file format. We just need to extend the comment parser. New versions of X584 will be perfectly able to open files produced by old versions. It may happen that some comments in old files inadvertently contain an `input` word. In this case, X584 will (mis)interpet it as the control directive. I guess it's not a big deal. We can happily live with it.

> If this extension will be implemented we won't be able to properly comment these sections: for example, we can't add extra comments inside conditional ones

Yeah, currently we cannot combine both control flow statements and ordinary comments in one line. This is a long-standing request from students. So I would suggest to first implement this feature in separate PR before adding an `input` keyword. For instance, we can split old comment column into two separate columns:

![x584](https://github.com/user-attachments/assets/c8fbc4fe-c61c-4805-9274-89bf31e6c3db)

To keep the backward compatibility with old `.x584` format, we can store both control flow statements and ordinary comments in the existing comment field but separate them with any reserved character, such as `\n`. New versions of X584 could detect this character and split the comment into two parts. This way, comments from the screenshot above would be stored in `.x584` file like the following:
```
input 0000 0000 0000 0001\nВводим мантиссу A
input 0000 0000 0000 0010\nВводим порядок A
if П then 3 else 5\nПроверяем П
```
Does this sound good?

**Comment by danil-kondr2016** [comment]
Yes, this sounds good.
I have edited comment inside Comment section (replaced one byte with '\n') and it renders as two joined with no symbol strings.
![image](https://github.com/user-attachments/assets/4e0c42cb-537c-438f-aa9d-374e23ddcd69)

But there are one problem: if control operator is the first "line" of comment, how we can determine where comment is **comment** and comment is **control operator**? I propose parse each comment during program load and if it has been parsed as control-operator it IS control operator otherwise it is plain comment. But if comment starts with '\n' then it is new format comment.

Another problem is length restriction: in the X584 file format maximal length of comment is 255 chars in CP1251. I propose 127 chars to control-operator and 127 for comment: 127+127+1=255 (but maybe we can restrict control operator with less bytes since they're usually has fixed length)

**Comment by danil-kondr2016** [comment]
I tested to copy comment and edit it (but not deleting symbol between "Control operator" and "Comment" and this byte has not been removed so we can edit it safely.

**Comment by danil-kondr2016** [comment]
I think this change will lead to increment major version of X584 (v2.00)

**Comment by danil-kondr2016** [comment]
Oh, I have another proposition of comment extending for v2.00:
English variants of register names! We have English **keywords** such as `if`, `then`, `else`, `goto` but we don't have **register names** in English!

|Russian|English|
|----|----|
|П|C|
|ПАЛУ0|C0|
|ПАЛУ1|C1|
|ПАЛУ2|С2|
|ПАЛУ3|C3|
|!СДЛ1|!SHL1|
|!СДП1|!SHR1|
|!СДЛ2|!SHL2|
|!СДП2|!SHR2|
|РРР0|XWR0|
|РРР3|XWR3|

**Comment by kodemeister** [comment]
Let's use this issue only for discussing the `input` directive. I've created separate issues for other proposed features:

* #21 
* #22 

To keep things clean, each PR should address exactly the single issue. PRs for issues marked with `[v2.0]` should go into `2.0` branch. I plan to keep legacy 1.x versions in `main` branch only for bugfixes and small non user-facing changes. New features should be developed for modern 2.x versions.

**Comment by danil-kondr2016** [comment]
So there is another problem. Which format will be used to specify input numbers? I propose four formats:

- binary format - 16 (without separators) or 19 (with separators) symbols long with only `'0'` and `'1'` symbols;
- decimal unsigned format - 1 to 5 symbols in range from `'0'` to `'9'` and specifying number at most of 65535;
- negative decimal - begins with `-` then 1 to 5 symbols in range from `'0'` to `'9'` and specifying number from `-1` to `-32768` in two's complement (should we support one's complement?)
- hexadecimal - begins with 0x and 1 to 4 symbols in `[0-9A-Fa-f]`

If we'll support only one format, this will not be handy for end-users

**Comment by kodemeister** [comment]
I agree. No need to support one's complement. While implementing this, please put all parsing code to separate method like `ParseNumber` instead of adding everything to already bloated `ParseComment`.

**Comment by kodemeister** [comment]
Implemented in #38.
